zio-http/src/main/scala/zio/http/Response.scala

```diff
@@ -77,44 +77,40 @@ final case class Response private (
    */
   def withServerTime: Response = self.copy(attribute = self.attribute.withServerTime)
 
+  private[zio] def withEncodedResponse(updated: Option[Response.EncodedResponse]): Unit = {
+    var loop = true
+    while (loop) {
+      val current = self.encodedResponse.get()
+      loop = !encodedResponse.compareAndSet(current, updated)
+    }
+  }
+
   private[zio] def close(implicit trace: Trace): Task[Unit] = self.attribute.channel match {
     case Some(channel) => NettyFutureExecutor.executed(channel.close())
     case None          => ZIO.refailCause(Cause.fail(new IOException("Channel context isn't available")))
   }
 
-  /**
-   * Encodes the Response into a Netty HttpResponse. Sets default headers such
-   * as `content-length`. For performance reasons, it is possible that it uses a
-   * FullHttpResponse if the complete data is available. Otherwise, it would
-   * create a DefaultHttpResponse without any content.
-   */
-  private[zio] def encode()(implicit trace: Trace): Task[HttpResponse] = for {
-    content <- if (body.isComplete) body.asChunk.map(Some(_)) else ZIO.succeed(None)
-    res     <-
-      ZIO.attempt {
-        import io.netty.handler.codec.http._
-        val jHeaders         = self.headers.encode
-        val hasContentLength = jHeaders.contains(HttpHeaderNames.CONTENT_LENGTH)
-        content.map(chunks => Unpooled.wrappedBuffer(chunks.toArray)) match {
-          case Some(jContent) =>
-            val jResponse = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, self.status.asJava, jContent, false)
-
-            // TODO: Unit test for this
-            // Client can't handle chunked responses and currently treats them as a FullHttpResponse.
-            // Due to this client limitation it is not possible to write a unit-test for this.
-            // Alternative would be to use sttp client for this use-case.
-            if (!hasContentLength) jHeaders.set(HttpHeaderNames.CONTENT_LENGTH, jContent.readableBytes())
-            jResponse.headers().add(jHeaders)
-            jResponse
-          case None           =>
-            if (!hasContentLength) jHeaders.set(HttpHeaderNames.TRANSFER_ENCODING, HttpHeaderValues.CHUNKED)
-            new DefaultHttpResponse(HttpVersion.HTTP_1_1, self.status.asJava, jHeaders)
-        }
-      }
-  } yield res
+  def copy(
+    status: Status = self.status,
+    headers: Headers = self.headers,
+    body: Body = self.body,
+    frozen: Boolean = self.frozen,
+    attribute: Response.Attribute = self.attribute,
+    encodedResponse: AtomicReference[Option[Response.EncodedResponse]] = self.encodedResponse,
```

Comment:
Because the mutable data is copied along, it is easy to violate referential transparency. Another reason to delete these from `Response`.

Reply:
See newest PoC approach using `ConcurrentMap`.